### PR TITLE
Add qiboml integration for QAOA in QUBO

### PR DESCRIPTION
**Pull request by jf-kong**
## Summary
This PR integrates `qiboml` as an optional training engine for `QUBO.train_QAOA`, while keeping the existing qibo optimizer path as default (`engine="legacy"`).

The goal is to enable gradient-based QAOA optimization through qiboml/torch with minimal API disruption and optional dependency installation.

## What Changed

1. Added qiboml adapter integration
- New module: `src/qiboopt/integrations/qiboml_adapter.py`
- Implements `optimize_qaoa_with_qiboml(...)` using qiboml's PyTorch `QuantumModel`
- Supports differentiation backends (`psr`, `jax`, `adjoint`, default torch path)
- Supports qiboml optimizers: `adam`, `sgd`
- Applies QUBO/Ising constant energy shift so reported losses match QUBO objective scale

2. Extended `QUBO.train_QAOA` to support multi-engine training
- New args: `engine`, `optimizer`, `lr`, `epochs`, `differentiation`
- `engine="legacy"` remains default and preserves previous behavior
- `engine="qiboml"` uses qiboml adapter for regular-loss optimization
- CVaR (`regular_loss=False`) continues through legacy path

3. Added exact (no-shot) execution support
- `nshots=None` or `nshots=0` now runs exact mode
- Return statistics are now:
  - sampled mode (`nshots > 0`): bitstring counts
  - exact mode (`nshots is None or 0`): bitstring probabilities

4. Refactored QAOA circuit construction utilities
- Added helpers for parameter splitting/rebuilding and callable circuit generation for qiboml tracing
- Added option to build circuits without measurements when needed by exact expectation paths

5. Dependency updates for optional qiboml usage
- `pyproject.toml` updated with optional `qiboml` and `torch`
- Added extras/group install path for qiboml runtime

## Backward Compatibility
- Default behavior remains unchanged (`engine="legacy"`).
- qiboml is fully optional and only required when `engine="qiboml"` is selected.

### CODE REVIEW

**Review by shangtai — COMMENTED** [review]
Thanks for this.  updated poetry.lock. 

We will have to fix some bugs first :) 

************* Module qiboopt.integrations.qiboml_adapter
src/qiboopt/integrations/qiboml_adapter.py:19:4: E0401: Unable to import 'qiboml.operations.differentiation' (import-error)
************* Module qiboopt.opt_class.opt_class
src/qiboopt/opt_class/opt_class.py:838:23: E0606: Possibly using variable 'params' before assignment (possibly-used-before-assignment)
src/qiboopt/opt_class/opt_class.py:857:16: E0606: Possibly using variable 'best' before assignment (possibly-used-before-assignment)
src/qiboopt/opt_class/opt_class.py:859:16: E0606: Possibly using variable 'extra' before assignment (possibly-used-before-assignment)
Error: Process completed with exit code 2.

**Review by shangtai — APPROVED** [review]

File: src/qiboopt/opt_class/opt_class.py [hunk 1]
```diff
@@ -599,22 +726,21 @@ def myloss(parameters):
...
+                    custom_mixer=custom_mixer,
+                    include_measurements=not use_exact,
+                    has_alphas=has_alphas,
                 )
                 if noise_model is not None:
                     circuit = noise_model.apply(circuit)
 
-                result = circuit(None, nshots)
+                if use_exact:
+                    hamiltonian = self.qubo_to_qaoa_object().hamiltonian
```
Comment (scarrazza) [inline src/qiboopt/opt_class/opt_class.py]: This call will cause O(maxiter) deterioration, which seems unnecessary. You can move the hamiltonian object allocation outside myloss.

File: src/qiboopt/integrations/qiboml_adapter.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        loss = model()
+        if loss.ndim > 0:
+            loss = loss.reshape(-1)[0]
+        loss.backward()
+        torch_optimizer.step()
+        loss_value = float(loss.detach().cpu().item()) + energy_shift
+        losses.append(loss_value)
+        if loss_value < best:
+            best = loss_value
+            current_parameters = model.circuit_parameters
```
Comment (scarrazza) [inline src/qiboopt/integrations/qiboml_adapter.py:141]: `model.circuit_parameters` is read after `torch_optimizer.step()`, but `loss_value` was computed from the parameters before the step. This means `best_params` may not correspond to the loss that triggered the best-tracking condition, leading to suboptimal or inconsistent final parameters.
Probably, this can be fixed by moving this line and the if condition belofre the loss.backward call.

File: src/qiboopt/integrations/qiboml_adapter.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    model = QuantumModel(
+        circuit_structure=[circuit_builder],
+        decoding=decoder,
+        parameters_initialization=np.asarray(parameters, dtype=np.float64),
+        differentiation=diff_class,
+    )
+    model = model.to(dtype=torch.float64)
+
+    if isinstance(optimizer, str):
+        _OPT_MAP = {
```
Comment (scarrazza) [inline src/qiboopt/integrations/qiboml_adapter.py]: Minor detail, but snake case here might be misleading, you simply use `opt_map`.

File: src/qiboopt/__init__.py [hunk 4]
```diff
@@ -1,5 +1,8 @@
-import importlib.metadata as im
+try:
+    import importlib.metadata as im
```
Comment (scarrazza) [inline src/qiboopt/__init__.py:2]: Why the try? You should use it and avoid bare except Exception:
```python
import importlib.metadata as im

try:
    __version__ = im.version(__package__)
except im.PackageNotFoundError:
    __version__ = "0.0.1"
```

File: doc/source/getting-started/quickstart.rst [hunk 5]
```diff
@@ -59,3 +59,31 @@ The Conditional Variance at Risk (CVaR) can also be used as an alternative loss
    gammas = [0.1, 0.2]
    betas = [0.3, 0.4]
    output = qp.train_QAOA(gammas=gammas, betas=betas, regular_loss=False, cvar_delta=0.1)
+
+To use qiboml's pytorch training loop instead of the legacy optimizer, set ``engine="qiboml"``:
+
+.. code-block:: python
+
+   from qiboopt.opt_class.opt_class import QUBO
```
Comment (scarrazza) [inline doc/source/getting-started/quickstart.rst:67]:
```suggestion
   from qiboopt.opt_class.opt_class import QUBO
   Qdict = {(0, 0): 1.0, (0, 1): 0.5, (1, 1): -1.0}
   qp = QUBO(0, Qdict)
```

File: src/qiboopt/__init__.py [hunk 6]
```diff
@@ -1,5 +1,8 @@
...
 
-__version__ = im.version(__package__)
+    __version__ = im.version(__package__)
+except Exception:
+    __version__ = "0.0.1"
 
-from qiboopt import combinatorial, opt_class
+from . import combinatorial, integrations, opt_class
```
Comment (scarrazza) [inline src/qiboopt/__init__.py:8]: I think you have to add a __init__.py inside the integrations folder otherwise this might generate a ModuleNotFoundError when doing import qiboopt.

File: src/qiboopt/opt_class/opt_class.py [hunk 7]
```diff
@@ -582,10 +677,42 @@ def train_QAOA(
...
+                "Falling back to engine='legacy'.",
+                UserWarning,
+                stacklevel=2,
+            )
+            engine = "legacy"
+        if not regular_loss and not (0 < cvar_delta <= 1):
+            raise_error(
+                ValueError,
+                f"cvar_delta must satisfy 0 < cvar_delta <= 1, but got {cvar_delta}.",
+            )
```
Comment (scarrazza) [inline src/qiboopt/opt_class/opt_class.py]: I would move this before engine fallback logic.

File: src/qiboopt/integrations/qiboml_adapter.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        raise ImportError(
+            "engine='qiboml' differentiation backend requires "
+            "`qiboml.operations.differentiation`."
+        ) from exc
+
+    mapping = {
+        "psr": differentiation_module.PSR,
+        "jax": differentiation_module.Jax,
+        "adjoint": differentiation_module.Adjoint,
+        "torch": None,
```
Comment (scarrazza) [inline src/qiboopt/integrations/qiboml_adapter.py]: For the sake of maintanability, I would suggest to remove torch from here given that it is handled by an early return.

File: src/qiboopt/opt_class/opt_class.py [hunk 9]
```diff
@@ -686,40 +815,54 @@ def myloss(parameters, delta=cvar_delta):
...
-        best, params, extra = optimize(
-            myloss, parameters, method=method, options={"maxiter": maxiter}
-        )
-        # Unpack optimised parameters in the same way as in myloss (block format)
-        if alphas is not None:
-            optimised_gammas = params[:p]
-            optimised_betas = params[p : 2 * p]
-            optimised_alphas = params[2 * p : 3 * p]
+        if engine == "qiboml":
+            best, params, extra = optimize_qaoa_with_qiboml(
```
Comment (scarrazza) [inline src/qiboopt/opt_class/opt_class.py:821]: Importing optimize_qaoa_with_qiboml at module load time forces the integrations package to be importable even when qiboml is never used. A lazy import inside the if engine == "qiboml" block would be more consistent with how the adapter itself lazily imports torch/qiboml.


```suggestion
            from qiboopt.integrations.qiboml_adapter import optimize_qaoa_with_qiboml
            best, params, extra = optimize_qaoa_with_qiboml(
```

File: src/qiboopt/integrations/qiboml_adapter.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    torch_optimizer = opt_cls(model.parameters(), lr=lr)
+
+    losses = []
+    best = float("inf")
+    best_params = np.asarray(parameters, dtype=np.float64)
+    for _ in range(epochs):
+        torch_optimizer.zero_grad()
+        loss = model()
+        if loss.ndim > 0:
+            loss = loss.reshape(-1)[0]
```
Comment (scarrazza) [inline src/qiboopt/integrations/qiboml_adapter.py]:
```suggestion
            loss = loss.squeeze()
```

File: src/qiboopt/integrations/qiboml_adapter.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+            "sgd": torch.optim.SGD,
+        }
+        opt_cls = _OPT_MAP.get(optimizer.lower())
+        if opt_cls is None:
+            raise ValueError(
+                f"Unknown optimizer string '{optimizer}'. "
+                f"Pass a torch.optim.Optimizer subclass directly, or use one of: {list(_OPT_MAP)}."
+            )
+    else:
+        opt_cls = optimizer  # user supplied a class directly
```
Comment (scarrazza) [inline src/qiboopt/integrations/qiboml_adapter.py:125]:
```suggestion
        if not callable(optimizer):
            raise TypeError(f"optimizer must be a string or callable, got {type(optimizer).__name__}")
        opt_cls = optimizer  # user supplied a class directly
```

File: tests/test_qiboml_integration.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        optimizer="adam",
+        lr=0.05,
+        epochs=30,
+    )
+    assert np.linalg.norm(params - init_params) > 0
+    assert np.isfinite(best)
+
+
+@pytest.mark.skipif(not _qiboml_available(), reason="qiboml/torch not installed")
+def test_qiboml_engine_supports_exact_mode():
```
Comment (scarrazza) [inline tests/test_qiboml_integration.py]:
```suggestion
@pytest.mark.parametrize("nshots", [None, 0])
def test_qiboml_engine_supports_exact_mode(nshots):
```

File: .github/workflows/rules.yml [hunk 13]
```diff
@@ -20,5 +20,5 @@ jobs:
       os: ${{ matrix.os }}
       python-version: ${{ matrix.python-version }}
       doctests: ${{ matrix.os == 'ubuntu-latest'}}
-      poetry-extras: "--with tests,docs"
+      poetry-extras: ${{ matrix.os == 'ubuntu-latest' && '--with tests,docs,qiboml' || '--with tests,docs' }}
```
Comment (scarrazza) [inline .github/workflows/rules.yml:23]: Do you really need that?

**Review by scarrazza — COMMENTED** [review]
Thanks @jf-kong for the effort, this looks like an excellent refactoring, please find some minor comments below.

File: src/qiboopt/__init__.py (hunk 6 above)
Comment (jf-kong) [reply src/qiboopt/__init__.py:8]: added init.py inside integrations folder

File: doc/source/getting-started/quickstart.rst (hunk 5 above)
Comment (jf-kong) [reply doc/source/getting-started/quickstart.rst:67]: added definition of `qp`

**Review by scarrazza — APPROVED** [review]

File: src/qiboopt/opt_class/opt_class.py (hunk 7 above)
Comment (jf-kong) [reply src/qiboopt/opt_class/opt_class.py]: Moved one block earlier.

File: src/qiboopt/integrations/qiboml_adapter.py (hunk 8 above)
Comment (jf-kong) [reply src/qiboopt/integrations/qiboml_adapter.py]: Removed torch

File: src/qiboopt/opt_class/opt_class.py (hunk 9 above)
Comment (jf-kong) [reply src/qiboopt/opt_class/opt_class.py:821]: thanks for the suggestion! change to lazy import.

File: src/qiboopt/integrations/qiboml_adapter.py (hunk 10 above)
Comment (jf-kong) [reply src/qiboopt/integrations/qiboml_adapter.py]: thanks for the suggestion! done.

File: src/qiboopt/integrations/qiboml_adapter.py (hunk 11 above)
Comment (jf-kong) [reply src/qiboopt/integrations/qiboml_adapter.py:125]: made the suggested change.
